## AI Programming Foundations Project

## Name: Sayed Saidoo

## Dataset: Wine(Chemical properties and quality scores)

## Project Overview

This project demonstrates a reproducible data analysis workflow using the Wine Quality datasets for Portuguese "Vinho Verde" red and white wines. The workflow includes data ingestion, cleaning, exploratory analysis, visualization, and interpretation using Python, Pandas, Numpy, Matplotlib, and Seaborn.

**The objective is to explore how physicochemical properties relate to expert-rated wine quality and to compare patterns between red and white wines.**

### Setup Core Dependencies and Data Ingestion

In [30]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Environment Ready")

Environment Ready


In [31]:
df_red = pd.read_csv("../data/winequality-red.csv", sep=";")
df_red.sample(7)

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
9,7.5,0.500,0.36,6.1,0.071,17.0,102.0,0.9978,3.35,0.80,10.5,5
370,6.9,0.765,0.02,2.3,0.063,35.0,63.0,0.9975,3.57,0.78,9.9,5
196,7.3,0.580,0.30,2.4,0.074,15.0,55.0,0.9968,3.46,0.59,10.2,5
1251,7.5,0.580,0.14,2.2,0.077,27.0,60.0,0.9963,3.28,0.59,9.8,5
587,7.3,0.730,0.24,1.9,0.108,18.0,102.0,0.9967,3.26,0.59,9.3,5
551,9.4,0.430,0.24,2.8,0.092,14.0,45.0,0.9980,3.19,0.73,10.0,6
1135,8.0,0.250,0.43,1.7,0.067,22.0,50.0,0.9946,3.38,0.60,11.9,6


In [32]:
df_white = pd.read_csv("../data/winequality-white.csv", sep=";")
df_white.sample(7)

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality
267,5.3,0.580,0.07,6.9,0.043,34.0,149.0,0.99440,3.34,0.57,9.7,5
3303,6.3,0.220,0.34,5.0,0.032,36.0,93.0,0.99012,3.27,0.36,13.5,7
442,6.0,0.250,0.28,2.2,0.026,54.0,126.0,0.98980,3.43,0.65,12.9,8
3741,7.0,0.150,0.28,14.7,0.051,29.0,149.0,0.99792,2.96,0.39,9.0,7
2311,7.0,0.220,0.26,2.8,0.036,44.0,132.0,0.99078,3.34,0.41,12.0,7
46,6.2,0.450,0.26,4.4,0.063,63.0,206.0,0.99400,3.27,0.52,9.8,4
1120,6.7,0.105,0.32,12.4,0.051,34.0,106.0,0.99800,3.54,0.45,9.2,6


### Data Inspection and Cleaning

In [33]:
print(df_red.shape)
print(df_white.shape)

(1599, 12)
(4898, 12)


In [34]:
df_red.info()

<class 'pandas.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [35]:
df_white.info()

<class 'pandas.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


- Initial inspection of both datasets showed that all columns were assigned appropriate numeric data types and no missing values were detected. This suggests the datasets are relatively clean and suitable for analysis, though additional validation checks such as duplicate detection and consistency checks will still be performed.

In [36]:
# Check for duplicate rows
print(f"Red wine duplicates: {df_red.duplicated().sum()}")
print(f"White wine duplicates: {df_white.duplicated().sum()}")

# Verify column names match
print(df_red.columns.equals(df_white.columns))

# Verify data types match
print(df_red.dtypes.equals(df_white.dtypes))

Red wine duplicates: 240
White wine duplicates: 937
True
True


In [37]:
duplicates_red = df_red[df_red.duplicated()]
duplicates_white = df_white[df_white.duplicated()]

In [ ]:
duplicates_red.sample(5)

In [ ]:
duplicates_white.sample(5)

- Duplicate rows were identified during the initial inspection and will be removed to improve data quality and prevent repeated observations from influencing the analysis.

- A consistency check was performed prior to merging the datasets. Both datasets contain the same columns and data types, indicating they can be combined safely without requiring structural modifications.

- The consistency check suggests that the red and white wine datasets can be merged into a single dataset for analysis. Adding a wine type indicator will allow comparisons between wine classes while preserving information about the original source of each observation.

### Cleaning

In [ ]:
def remove_duplicates(df: pd.DataFrame) -> pd.DataFrame:
    """
    Remove duplicate rows from a dataframe.
    
    Parameters:
    df (pd.DataFrame): Input dataframe.
    
    Returns:
    pd.DataFrame: Dataframe with duplicate rows removed.
    
    """
    return df.drop_duplicates()

In [ ]:
df_red = remove_duplicates(df_red)
df_white = remove_duplicates(df_white)

In [ ]:
print(df_red.shape)
print(df_white.shape)

In [ ]:
def add_wine_type(df: pd.DataFrame, wine_type: str) -> pd.DataFrame:
    """
    Add a wine type identifier to a dataframe.
    
    Parameters:
    
        df (pd.DataFrame): Input dataframe.
        wine_type (str): Label indicating the wine type.
    
    Returns:
        pd.DataFrame: Dataframe with a new wine_type column.
    
    """
    df = df.copy()
    df["wine_type"] = wine_type
    return df

In [ ]:
df_red = add_wine_type(df_red, "Red")
df_white = add_wine_type(df_white, "White")

In [ ]:
df_red.head()

In [ ]:
df_white.head()

- A wine_type column was added prior to merging the datasets so that observations from red and white wines could be distinguished after consolidation.
This preserves the origin of each record and enables comparative analysis between wine classes.

In [ ]:
def merge_wine_datasets(df_red: pd.DataFrame, df_white: pd.DataFrame) -> pd.DataFrame:
    """
    Merge the red and white wine datasets into a single dataframe.
    
    Parameters:
        df_red (pd.DataFrame): Cleaned red wine dataframe.
        df_white (pd.DataFrame): Cleaned white wine dataframe.
    
    Returns:
        pd.DataFrame: Combined wine dataframe.
    """
    return pd.concat([df_red, df_white], ignore_index=True)

In [ ]:
df_wine = merge_wine_datasets(df_red, df_white)

In [ ]:
# Confirmation of merge success
df_wine.head()

In [ ]:
# Sanity check for white wine rows
df_wine.tail()

### Exploratory Data Analysis

In [ ]:
# Summary Statistics
def summary_statistics(df: pd.DataFrame) -> pd.DataFrame:
    """
    Generate summary statistics for the wine dataset.
    
    Parameters:
    
    df (pd.DataFrame): Input dataframe.
    
    Returns:
    
    pd.DataFrame: Summary statistics for numerical features.
    """
    return df.describe()
summary_statistics(df_wine)

**Interpretation**

 - The combined dataset contains 5,320 wine observations with no missing values.

 - Several variables exhibit substantial variability, particularly residual sugar, free sulfur dioxide, and total sulfur dioxide, as indicated by their relatively large standard deviations and wide ranges. This suggests meaningful differences between wine samples and potential opportunities for further investigation.

- Wine quality scores range from 3 to 9, with a median and 75th quartile settling at 6, indicating that most wines are concentrated around average quality ratings. This suggests the dataset contains fewer low and high-quality wines, which may introduce class imbalance considerations in future predictive modeling tasks if project is extended into predicitve modeling.

- The difference between the mean (5.80) and median (6.00) quality score is small, suggesting the quality distribution is not heavily skewed.

- The presence of large differences between minimum and maximum values for several variables suggests potential outliers or naturally occurring variability within the wine production process. These observations may warrant further examination through visualization and correlation analysis.

In [ ]:
# Looking at average feature values by wine type
def wine_type_means(df: pd.DataFrame) -> pd.DataFrame:
    """
    Generate average feature values grouped by wine type.
    
    Parameters:
    df (pd.DataFrame): Combined wine dataset.
    
    Returns:
    pd.DataFrame: Mean values of numerical features grouped by wine type.
    """
    return df.groupby("wine_type").mean(numeric_only=True)
wine_type_means(df_wine)

**Interpretation**

- Both wine types received similar average quality ratings. Red wines achieved an average quality score of 5.62, while white wines achieved an average quality score of 5.85, suggesting that expert ratings are broadly comparable across both wine categories.

- White wines contain substantially higher average levels of residual sugar, free and total sulfur dioxide than red wines. These differences may reflect distinct production and preservation processes between the two wine types or just a difference in the chemical composition of raw products themselves.

- Red wines exhibit higher average values for fixed acidity, volatile acidity, and sulphates, indicating notable differences in their physicochemical composition compared to white wines.

- Despite these chemical differences, the similarity in average quality scores suggests that wine quality is likely influenced by multiple interacting factors rather than a single physicochemical characteristic.

- As a next step it would be interesting to see if there are any correlations/interactions between features (dependent/independent)

In [ ]:
df_wine.corr(numeric_only=True)["quality"].sort_values(ascending=False)

**Interpretation**

- Alcohol exhibited the strongest positive relationship with quality (0.47), suggesting that wines with higher alcohol content tended to receive higher expert quality ratings.

- Density showed the strongest negative relationship with quality (-0.33), while volatile acidity (-0.27) and chlorides (-0.20) also demonstrated moderate negative relationships. These characteristics may be associated with lower-rated wines.

- Most remaining variables exhibited relatively weak correlations with quality, perhaps reinforcing early indications that wine quality is likely influenced by a combination of factors rather than a single physicochemical measurement.

- While correlation does not imply causation, these results suggest that alcohol, density, volatile acidity, and chlorides may be the most informative variables for understanding wine quality within this dataset.

### Visualizations

In [ ]:
# We will look at the Distribution or Wine Quality by Wine Type, given the sample imbalance we will create a subplot for each wine type.
sns.set_theme(style="whitegrid")
RED = "#D55E00" # colors that take color blindness into account
BLUE = "#0072B2"

fig, axes = plt.subplots(1, 2, figsize=(12, 5), sharey=True)

# Red Wine
sns.countplot(data=df_wine[df_wine["wine_type"] == "Red"], x="quality", color=RED, ax=axes[0])
axes[0].set_title("Red Wine Quality Distribution")
axes[0].set_xlabel("Quality Score")
axes[0].set_ylabel("Count")

# White Wine
sns.countplot(data=df_wine[df_wine["wine_type"] == "White"], x="quality", color=BLUE, ax=axes[1])
axes[1].set_title("White Wine Quality Distribution")
axes[1].set_xlabel("Quality Score")
axes[1].set_ylabel("Count")

plt.suptitle("Distribution of Wine Quality by Wine Type", fontsize=14)
plt.tight_layout()
plt.savefig("../images/figure_1_quality_distribution.png",dpi=300,bbox_inches="tight")
plt.show()

**Interpretation**

- Both red and white wines are concentrated around quality scores of 5 and 6, indicating that most wines in the dataset received average expert ratings, reinforcing our previous EDA. White wines exhibit a slightly higher concentration of quality scores above 6, which is consistent with the grouped analysis showing a marginally higher average quality rating for white wines. High-quality and low-quality wines are relatively uncommon in both groups, suggesting that extreme ratings are underrepresented within the dataset.

In [ ]:
palette = {
"Red": "#D55E00",
"White": "#0072B2"
}
plt.figure(figsize=(10, 6))
sns.boxplot(data=df_wine, x="quality", y="alcohol", hue="wine_type", palette=palette)
plt.title("Alcohol Content by Quality Score and Wine Type")
plt.xlabel("Quality Score")
plt.ylabel("Alcohol (%)")
plt.savefig("../images/figure_2_alcohol_content_quality_score.png",dpi=300,bbox_inches="tight")
plt.show()

**Interpretation**

- A clear positive relationship is observed between alcohol content and wine quality. Median alcohol levels generally increase as quality scores rise, particularly for wines rated 6 and above. This finding is consistent with the correlation analysis, which identified alcohol as the strongest positive correlate of quality.

- While the overall trend is upward, several outliers are present. Notably, some wines with a quality score of 5 exhibit relatively high alcohol content, suggesting that alcohol alone is insufficient to explain expert quality ratings. This reinforces the earlier observation that wine quality is likely influenced by multiple interacting physicochemical characteristics rather than a single dominant factor.

- An additional observation is that no red wines received the highest quality score of 9 in the dataset, while a small number of white wines achieved this rating. Although the sample of highly rated wines is small, this may partially explain the slightly higher average quality observed among white wines.

In [ ]:
# Density by Quality score by wine type
plt.figure(figsize=(10, 6))
sns.boxplot(data=df_wine, x="quality", y="density", hue="wine_type",showfliers=False, palette=palette)
plt.title("Density by Quality Score and Wine Type")
plt.xlabel("Quality Score")
plt.ylabel("Density")
plt.savefig("../images/figure_3_density_quality_score.png",dpi=300,bbox_inches="tight")
plt.show()

**Notebook Note**

An extreme density observation was retained in the dataset but omitted from the visual display using showfliers=False. This approach improves readability by allowing the central distribution of density values to be examined more clearly while preserving the original data for analysis.

**Interpretation**

- The density distributions differ between red and white wines, reflecting the compositional differences identified during exploratory analysis. Despite these differences, a consistent trend is observed across both wine types: higher-quality wines tend to exhibit lower density values, focusing on the median values here.

- The relationship becomes particularly noticeable for wines with quality scores above 6, where median density values generally decrease slightly as quality ratings increase. This finding is consistent with the correlation analysis, which identified density as the strongest negative correlate of wine quality in the combined dataset.

In [ ]:
# correlation plot for features
plt.figure(figsize=(12, 8))
sns.heatmap(df_wine.corr(numeric_only=True), cmap="coolwarm", annot=True, fmt=".2f")
plt.title("Correlation Heatmap of Wine Features")
plt.savefig("../images/figure_4_feature_correlation.png",dpi=300,bbox_inches="tight")
plt.show()

**Interpretation**

- The correlation heatmap provides an overall view of relationships between the physicochemical properties and wine quality ratings. Consistent with the earlier correlation analysis, alcohol content exhibits the strongest positive relationship with quality, while density shows the strongest negative relationship.

- Several relationships between predictor variables are also apparent. Free sulfur dioxide and total sulfur dioxide demonstrate a strong positive correlation (0.72) which aligns, while alcohol and density exhibit a strong negative correlation (-0.67). Moderate positive relationships are observed between residual sugar and density, as well as residual sugar and total sulfur dioxide.

- Despite these associations, most correlation coefficients remain below 0.60 in absolute value, indicating limited multicollinearity among the measured variables. This suggests that wine quality is influenced by multiple physicochemical properties acting together rather than a small number of highly correlated features.

## Summary and Overall Interpretations



**What I Learned from the Dataset**

This analysis explored the physicochemical properties and expert quality ratings of Portuguese "Vinho Verde" red and white wines. After removing duplicate records and merging the datasets, the final dataset contained 5,320 observations with no missing values, the dataset contained complete numerical measurements with appropriate data types and required only minimal cleaning prior to analysis. The analysis showed that red and white wines differ noticeably in their chemical composition, particularly in measures such as residual sugar, sulfur dioxide, acidity, and density. Several physicochemical features exhibited measurable relationships with wine quality but overall, quality was suggested to be influenced by multiple physicochemical properties acting together rather than a small number of highly correlated features.

**Interesting Patterns and Insights**

Alcohol content exhibited the strongest positive relationship with wine quality, while density showed the strongest negative relationship. Visualizations further demonstrated that higher-quality wines generally contained higher alcohol levels and lower density values, particularly among wines with quality scores above 6. The analysis also showed that white wines contained a small number of wines rated 9, while no red wines received the highest quality rating in the dataset.

The correlation analysis revealed relatively few strong relationships among the measured variables. While some features, such as free and total sulfur dioxide, displayed moderate to strong correlations, most variables exhibited only weak to moderate associations with one another and with wine quality.

**Limitations and Assumptions**

Several limitations should be considered when interpreting these results. The dataset contains only physicochemical measurements and expert quality ratings, with no information about grape variety, production methods, wine brand, price, region, or consumer preferences. As a result, potentially important factors influencing wine quality were not available for analysis.

In addition, quality scores were not evenly distributed across all rating categories. Most wines received average ratings of 5 or 6, while very high-quality and very low-quality wines were relatively uncommon. This distribution may limit the ability to draw strong conclusions about wines at the extremes of the quality scale.

Finally, the analysis focused on identifying patterns and associations rather than establishing causal relationships. Correlation results should therefore be interpreted as indicators of association rather than evidence of cause and effect.

**Surprising or Unclear Findings**

One of the most surprising findings was that red and white wines achieved similar average quality ratings despite exhibiting substantial differences in several chemical properties. It was also notable that alcohol content showed a stronger relationship with quality than many other measured characteristics.

At the same time, no single variable demonstrated an overwhelming relationship with quality. This suggests that expert wine ratings likely reflect a combination of interacting physicochemical factors rather than a single dominant attribute. Further analysis or predictive modeling would be required to better understand how these characteristics collectively influence wine quality.

**Bias and Data Quality Considerations**

The dataset uses expert assigned quality ratings as the target variable. While expert evaluation provides a structured assessment of wine quality, these ratings may still reflect subjective preferences and may not fully represent broader consumer perceptions of wine quality. In addition, the dataset contains substantially more average-quality wines than very high- or very low-quality wines, which may influence the conclusions drawn from the analysis and could introduce bias if the dataset were later used for predictive modeling.